ambry-store/src/main/java/com/github/ambry/store/IndexSegment.java

```diff
@@ -838,8 +848,10 @@ void seal() throws StoreException {
    * @return index value of last PUT record in this index segment. Return {@code null} if no PUT is found
    */
   IndexValue getIndexValueOfLastPut() throws StoreException {
+
     IndexValue indexValueOfLastPut = null;
-    if (sealed.get()) {
+    NavigableMap<StoreKey, ConcurrentSkipListSet<IndexValue>> indexCopy = index;
+    if (isSealedLocked()) {
```

Comment:
Same as above. You execute the critical path without holding the lock. You need to ensure isSealed is consistent throughout the execution of this routine. So, you need to get the lock in the beginning of this routine and hold it until you return.